File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int n, total;
    int dp[1001][2001];
    int solve(int idx, int sum, vector<int> &nums, int target) {
        if(idx >= n) return sum == target;

        if(dp[idx][sum + total] != -1) return dp[idx][sum + total];

        int add = solve(idx + 1, sum + nums[idx], nums, target);
        int subtract = solve(idx + 1, sum - nums[idx], nums, target);

        return dp[idx][sum + total] = add + subtract;
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        this -> n = nums.size();
        this -> total = accumulate(nums.begin(), nums.end(), 0);

        memset(dp, -1, sizeof(dp));

        return solve(0, 0, nums, target);
    }
};

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        total = sum(nums)

        dp = [[-1] * (2001) for _ in range(1001)]

        def solve(idx, CurSum) ->int:
            if idx >= n:
                return CurSum == target

            if dp[idx][CurSum + total] != -1:
                return dp[idx][CurSum + total]

            sub = solve(idx + 1, CurSum - nums[idx])
            add = solve(idx + 1, CurSum + nums[idx])

            dp[idx][CurSum + total] = (add + sub)
            return dp[idx][CurSum + total]

        return solve(0, 0)